src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsExtensionBuilderExtensions.cs

```diff
@@ -69,9 +71,9 @@ public static class WebJobsExtensionBuilderExtensions
                 throw new ArgumentNullException(nameof(configure));
             }
 
-            builder.Services.AddSingleton<IConfigureOptions<TOptions>>(p =>
-                new WebJobsExtensionOptionsConfiguration<TOptions>(p.GetService<IConfiguration>(), builder.ExtensionInfo.ConfigurationSectionName, configure));
-
+            builder.Services.AddSingleton<IConfigureOptions<TOptions>>((IServiceProvider p) => new WebJobsExtensionOptionsConfiguration<TOptions>(p.GetService<IConfiguration>(), builder.ExtensionInfo, configure));
```

Comment:
This will add multiple different instances to the container which isn't what we want. We want to register the same single instance and have that same instance be returned regardless of which interface it is requested through. There are already examples of how to do this, e.g. [here](https://github.com/Azure/azure-functions-host/blob/d49c3b94d89d0cec419b49ef22999fe6ae57d9c3/src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs#L159). In this example, WebJobsScriptHostService is registered as WebJobsScriptHostService, IHostedService, and later below as IScriptHostManager.

Reply:
Hi @mathewc  
I tried, TryAdd with non-generic interface strategy, however, it throws exception. TryAddEnumerable only allows ServiceDescriptor.   I tested two ways. 

```csharp
builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IWebJobsExtensionOptionsConfiguration>(p =>
            
      (IWebJobsExtensionOptionsConfiguration)p.GetRequiredService<IConfigureOptions<TOptions>>()
 ));
```
It causes 

```
System.ArgumentException : Implementation type cannot be 'Microsoft.Azure.WebJobs.Host.Config.IWebJobsExtensionOptionsConfiguration' because it is indistinguishable from other services registered for 'Microsoft.Azure.WebJobs.Host.Config.IWebJobsExtensionOptionsConfiguration'. (Parameter 'descriptor')
```
I found one issue on the dotnet repo. https://github.com/dotnet/runtime/issues/56105 I use the workaround as well, however, it doesn't work. Am I missing something?

Reply:
Hi @mathewc 
I tried the approach. TryAddEnumerable strategy works well. However, TryAddEnumerable + `p.GetRequiredService<IConfigureOptions<TOptions>()` approach doesn't work. 

## The `p.GetRequiredService` approach doesn't work
Using this approach, it works with testing on the WebJobs side, however, when I work from Azure Functions Host side, I see the following error. It looks confusing with the other configuration. It can be possible that my DI configuration on Azure Functions is wrong. 

```csharp
          builder.Services.AddSingleton<IConfigureOptions<TOptions>>(p => new WebJobsExtensionOptionsConfiguration<TOptions>(p.GetService<IConfiguration>(), builder.ExtensionInfo, configure));
            builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IWebJobsExtensionOptionsConfiguration, WebJobsExtensionOptionsConfiguration<TOptions>>(p => (WebJobsExtensionOptionsConfiguration<TOptions>)p.GetRequiredService<IConfigureOptions<TOptions>>()));
```

_error_

```
System.InvalidCastException: 'Unable to cast object of type 
'Microsoft.Extensions.Options.ConfigureNamedOptions`1[Microsoft.Azure.WebJobs.EventHubs.EventHubOptions]' to type 
'Microsoft.Azure.WebJobs.Host.Config.WebJobsExtensionOptionsConfiguration`1[Microsoft.Azure.WebJobs.EventHubs.EventHubOptions]'.'
```

## current approach works
However, I have duplicate instance. That the part I don't like. 

```csharp
           builder.Services.AddSingleton<IConfigureOptions<TOptions>>(p => new WebJobsExtensionOptionsConfiguration<TOptions>(p.GetService<IConfiguration>(), builder.ExtensionInfo, configure));
            builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IWebJobsExtensionOptionsConfiguration, WebJobsExtensionOptionsConfiguration<TOptions>>(p => new WebJobsExtensionOptionsConfiguration<TOptions>(p.GetService<IConfiguration>(), builder.ExtensionInfo, configure)));
```

Reply:
Done